src/compact_log.rs

```diff
@@ -119,40 +160,188 @@ pub(crate) fn split_next_segment(
+    type Item = LogicalClock;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.remaining_clock_region.len() < 2 {
+            return None;
+        }
+        let (curr, remainder) = self.remaining_clock_region.split_at(2);
+        self.remaining_clock_region = remainder;
+        Some(LogicalClock {
+            // N.B. Currently relying on the fact that these are internally-managed
+            // clocks to ensure their validity.
+            id: TracerId(unsafe {
+                NonZeroU32::new_unchecked(curr[0].interpret_as_logical_clock_tracer_id())
+            }),
+            count: curr[1].raw(),
+        })
+    }
+}
+
+/// Iterator for iterating through the events in a given compact log segment
+pub struct LogSegmentEventIterator<'a> {
+    remaining_event_region: &'a [CompactLogItem],
+}
+
+/// The sanity checks that might fail when interpreting the compact log format
+/// into a concrete stream of events.
+#[derive(Debug)]
+pub enum LogEventInterpretationError {
+    /// The event id was not in the allowed ekotrace::EventId ranges
+    InvalidEventId(u32),
+    /// An event with the "has a payload" flag was found, but the payload-bearing
+    /// log item expected next was not present. This may happen when iterating
+    /// over an incomplete or incorrectly segmented slice of compact log items.
+    EventMissingPayload(EventId),
+}
+
+impl<'a> Iterator for LogSegmentEventIterator<'a> {
+    type Item = Result<LogEvent, LogEventInterpretationError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some((first, remaining)) = self.remaining_event_region.split_first() {
+            self.remaining_event_region = remaining;
+            let raw_ev = first.raw();
+            if first.has_event_with_payload_bit_set() {
+                let event_expecting_payload = raw_ev & !EVENT_WITH_PAYLOAD_MASK;
+                let event_expecting_payload = event_expecting_payload.try_into().map_err(|_e| {
+                    LogEventInterpretationError::InvalidEventId(event_expecting_payload)
+                });
+                if let Some((second, remaining)) = self.remaining_event_region.split_first() {
+                    self.remaining_event_region = remaining;
+                    Some(
+                        event_expecting_payload
+                            .map(|event_id| LogEvent::EventWithPayload(event_id, second.raw())),
+                    )
+                } else {
+                    Some(match event_expecting_payload {
+                        Ok(event_id) => {
+                            Err(LogEventInterpretationError::EventMissingPayload(event_id))
+                        }
+                        Err(e) => Err(e),
+                    })
+                }
+            } else {
+                Some(
+                    raw_ev
+                        .try_into()
+                        .map_err(|_e| LogEventInterpretationError::InvalidEventId(raw_ev))
```

Comment:
Am I correct in thinking that this is effectively impossible to hit, due to the above bit test?

Reply:
(at line 279)

Reply:
Even if the "event with payload" bit is set, doesn't mean that the event is valid, so the prior check on line 279 is not a complete validity pass. For example, the clock bit could also have been set or the underlying raw value was 0.

That said, assuming that the log has been properly created and segmented, I don't expect the highlighted log interpretation errors to surface. The presence of a full validity check via try_into is probably a necessary-but-rare protection against incorrect construction or accidental regressions.